include/dynamatic/Dialect/Handshake/HandshakeArithOps.td

```diff
@@ -228,7 +228,10 @@ def Handshake_ConstantOp : Handshake_Arith_Op<"constant", [
 
   let arguments = (ins TypedAttrInterface:$value, ControlType:$ctrl);
   let results = (outs SimpleChannel:$result);
-  let assemblyFormat = "$ctrl attr-dict `:` type($result)";
+
+  // The type of the control also needs to be specified in the IR.
+  // It may have extra bits, which could affect the result's type and token.
+  let assemblyFormat = "$ctrl attr-dict `:` type($ctrl) `,` type($result)";
```

Comment:
I'll update the type of result to `ChannelType` in the next PR focusing on the spec tag

Reply:
Just make `$ctrl` a `SimpleControl` for now since the result is a `SimpleChannel`, instead of changing the assembly format.

Is the intent to forward the extra signals from `$ctrl` into `$result`? That makes the constant not so constant anymore. Doesn't it make more sense to have those extra signals be encoded in `$value`?

Reply:
It would simplify type-checking if constants in the speculative region had spec bits- the easiest way to implement this is to take the spec bit of the input token, and have source operations output tokens marked as non-spec. This would allows us to use all type match on basically all operations (with the exception of mux and cmerge).

Otherwise every unit that could ever take a constant input will need custom type verification implemented.

Reply:
> and have source operations output tokens marked as non-spec

So just a simple control as it is now? So will you need a SourceOp which outputs a control with extra signals? If so how do you define the content of extra signals at the source? My comment about constants becoming not really constant (in extra signals) remains as well. I am afraid that semantics are going to get messy. 

> Otherwise every unit that could ever take a constant input will need custom type verification implemented.

I really don't understand what you mean there.

Reply:
A simple control token is not a non-speculative token, in the same way a signless integer is not an unsigned integer.

If a constant supplies a value of 5 to an add operation, we believe it should also supply a 0 in the speculative channel to feed the logic that will generate an output for the speculative value. Otherwise, the add operation would have to adapt to some inputs being simple channels, which is not really the responsibility of the add operation. 

In any case, at some point this 0 value for the spec channel must be generated, whether explicitly or implicitly. The most natural place to generate it is from the source operation, as it means no other operations are forced to generate (or assume) speculative bit values.

>  Otherwise every unit that could ever take a constant input will need custom type verification implemented.
I really don't understand what you mean there.

The majority of the type verification is all types match. If some inputs are simple channels and some are not, we can no longer use all types match. By making source units in the speculative regions issue explicitly non-speculative tokens, it means every channel in the speculative region is the same type, allowing us to use all types match, and simplifing the rest of our code.

Reply:
I agree with all of that, in fact it is how I assumed people would use these operations in the first place. What I am asking is what makes the most sense between the source or the constant being responsible for generating that initial spec bit; I think there is a clear winner.

```mlir
// pseudocode
%val = ... : channel<i32, [spec: i1]> 

// Solution 1 - Source provides extra signal
%ctrl_with_extra = handshake.source {spec: i1} : <[spec: i1]>
%cst_with_extra = handshake.constant %ctrl_with_extra {value = <some-val>: i32} : channel<i32, [spec: i1]> 
%add = handshake.addi %val, %cst_with_extra : channel<i32, [spec: i1]>

// Solution 2 - Constant provides extra signal
%ctrl = handshake.source : <>
%cst_with_extra = handshake.constant %ctrl {value = {<some-val>: i32, [spec = <some-init-val>: i1]} : channel<i32, [spec: i1]> 
%add = handshake.addi %val, %cst_with_extra : channel<i32, [spec: i1]>
```

Solution 2 by far makes the most sense to me.

- The constant actually remains constant (not the case in solution 1, since extra signals from the input control could change at seemingly random times).
- No undefined behavior. In solution 1 we don't know what the value of `spec` is, so it's completely up in the air.
- Simpler op interfaces. Sources always produce a simple control and constants always take a simple control input, like we have now.
- Homogeneity. Solution 1 makes the content of data come from the constant and the content of extra signals from the input. The `value` attribute for constants should encode everything. We should just create an attribute that is able to express the initial values of data + an arbitrary list of extra signals (or undef if you don't care), and the result type of the op just matches that attribute, as it does now. You could even have the constant generate a control with extra signals i.e., data could be optional.

Reply:
My issue with 
`%cst_with_extra = handshake.constant %ctrl {value = {<some-val>: i32, [spec = <some-init-val>: i1]} : channel<i32, [spec: i1]> `
is the parameterizability of the init value for the spec tokens- the spec signal should always be non-spec.

I think the opposite of your point about undefined behavior, the IR should include information that is not uniquely decided by the operation name- source operations should only produce non-speculative tokens.

Your simpler op interface point is, unfortunately, false, as when the control path for a basic block is speculative (basically always) the constants produced by the control tokens will not receive a simple control input. Maybe this is avoidable, but I don't really understand why constants in Dynamatic are so complicated/have multiple formulations.  

From my point of view, it is simpler to have constants in the speculative region always receive speculative channels than sometimes. I also like the coherence of all channels in the speculative region having spec signals, rather than the edges from the sources being simple.

What I don't like about this solution is that sometimes constants are speculative (if produced by a control token), but I also don't think I like the constant receiving a speculative channel value that it ignores.

Reply:
I see your points, especially the one about consistency in speculative regions and the fact that you will naturally have controls with extra signals going around that you would like to use. The uses of constants with extra signals in the control input just confuses me, in part because it makes the constant just be a "tie data to this control with extra signals" operation (which seems more like a bundle kind of op, but as discussed elsewhere this has its own set of challenges), but I guess that's fine in the end. Feel free to proceed however you see fit. As long as the semantics of the op are clear I won't oppose it.

> I don't really understand why constants in Dynamatic are so complicated/have multiple formulations.

Do you mean the existence of both `arith.constant` and `handshake.constant`? If so we just can't use the former because it won't accept our custom types. Other than that the Handshake constant seems as straightforward as it can be to me; too much in fact since you can't do what you want with it currently.

> I also don't think I like the constant receiving a speculative channel value that it ignores.

Absolutely, I would not allow this in solution 1. To me the constant output always has the same extra signals as its input in this case.

Reply:
So I think we're more or less agreed :)

> Do you mean the existence of both arith.constant and handshake.constant?

More so that some constant ops are connected to source ops and some aren't, and the process of how that is decided.

> The uses of constants with extra signals in the control input just confuses me,
I find in general source and constant being separate operations confusing, as to why constants take any input at all?

Reply:
I’m aware of cases where separating source and constant ops is necessary to match the number of control tokens triggering the channel tokens (like the case in the data speculation benchmark). I think the details of this case go beyond our current discussion, so I’ll talk about them with you later in another place.

For this PR, I think it’s better to limit the control type to SimpleControl, as the implementation shouldn’t depend on another PR (considering the case where we need to revert another PR later). I’ll relax this restriction in the next PR, so I’ll update the argument type for now.

Thanks for raising the question about whether constant ops should handle extra signals. I hadn’t considered how it could confuse semantics (making them not truly "constant"), but you’re right. Using Bundle/Unbundle to extract SimpleControl simplifies the semantics of ConstantOp, but it does raise the question: why does only ConstantOp need Bundle/Unbundle, while other ops like Add handle extra signals internally?

Reply:
To be honest constants were designed at a time where the objective was "produce the same CDFG as legacy Dynamatic with new Dynamatic", so I am really not saying that their interface intrinsically makes a lot of sense; this discussion gives some good evidence that it does not.

The idea of the control is that you don't want constants to be always triggering because of the side-effects that may have on the circuit. If the C code contains a `mem[0] = 0;` and constants are always valid then the store will "trigger all the time". This was at least true in legacy but now that we are generally improving our interfaces perhaps this has become a non-issue. The decision to connect to a source or to the control network is driven by this very worry, and the logic to do it is arguably broken but "it works".

https://github.com/EPFL-LAP/dynamatic/blob/13948ba2710c961ed037d14d8eaac415692d416c/lib/Conversion/CfToHandshake/CfToHandshake.cpp#L1193-L1203

> but it does raise the question: why does only ConstantOp need Bundle/Unbundle, while other ops like Add handle extra signals internally?

I think the difference of importance here is that an adder "does not care" for extra signals, it just has to forward them somehow. The current "constant" can be seen as an op whose only point is to add "an extra signal" to its input, in the current implementation a data bus to a control, so it is definable as some combination of bundle/unbundle. The adder is not definable with bundle/unbundle, it actually does something else.